libraries/core_libs/consensus/src/transaction/transaction_queue.cpp

```diff
@@ -77,25 +104,45 @@ bool TransactionQueue::insert(std::shared_ptr<Transaction> &&transaction, const
 
   switch (status) {
     case TransactionStatus::Verified: {
-      const auto it = priority_queue_.insert(std::move(transaction));
-
-      // This assert is here to check if priorityComparator works correctly. If object is not inserted, then there could
-      // be something wrong with comparator
-      assert(it != priority_queue_.end());
-
-      // This check if priority_queue_ is not bigger than max size if so we delete last object
-      // if the last object is also current one we return false
-      if (priority_queue_.size() > kMaxSize) [[unlikely]] {
-        const auto last_el = std::prev(priority_queue_.end());
-        if (it == last_el) {
-          priority_queue_.erase(last_el);
+      const auto &account_it = account_nonce_transactions_.find(transaction->getSender());
+      if (account_it == account_nonce_transactions_.end()) {
+        account_nonce_transactions_[transaction->getSender()][transaction->getNonce()] = transaction;
+        queue_transactions_[tx_hash] = transaction;
+      } else {
+        const auto &nonce_it = account_it->second.find(transaction->getNonce());
+        if (nonce_it == account_it->second.end()) {
+          account_nonce_transactions_[transaction->getSender()][transaction->getNonce()] = transaction;
+          queue_transactions_[tx_hash] = transaction;
+        } else {
+          // It should not be possible that transaction is already inside due to verification done before
+          assert(nonce_it->second->getHash() != tx_hash);
+          // Replace transaction if gas price higher
+          if (transaction->getGasPrice() > nonce_it->second->getGasPrice()) {
+            // Place same nonce transaction with lower gas price in non propsable transactions since it could be
+            // possible that some dag block might contain it
+            non_proposable_transactions_[nonce_it->second->getHash()] = {last_block_number, nonce_it->second};
+            queue_transactions_.erase(nonce_it->second->getHash());
+            account_nonce_transactions_[transaction->getSender()][transaction->getNonce()] = transaction;
+            queue_transactions_[tx_hash] = transaction;
+          }
+        }
+      }
+
+      auto queue_size = size();
+      // This check if priority_queue_ is not bigger than max size if so we delete 1% of transactions
```

Comment:
I forgot why do we delete 1% == 2000 txs ?

Reply:
ordering is expensive so this way it is done every 2000 new transactions received and not for every transaction received if we are at the limit. Ordering is only done when:
1. We are proposing dag block
2. We are over the max size limit

Reply:
ok lets see how that behaves during stress tests